Add purge_tags to s3_bucket to allow preservation of existing tags

willthames opened the pull request:
##### SUMMARY
Adding `purge_tags` with default `True` to maintain existing behaviour
allows users to set it to `False` to preserve existing tags

Fixes #29366


##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
s3_bucket

willthames committed:
Add purge_tags to s3_bucket to allow preservation of existing tags

Adding `purge_tags` with default `True` to maintain existing behaviour
allows users to set it to `False` to preserve existing tags

Fixes #29366

s-hertel commented on test/integration/targets/s3_bucket/tasks/main.yml:
```diff
@@ -6,10 +6,10 @@
-          region: "{{ aws_region }}"
+          aws_access_key: "{{ aws_access_key | default('') }}"
+          aws_secret_key: "{{ aws_secret_key | default('') }}"
+          security_token: "{{ security_token | default('') }}"
```
This is convenient.

willthames replied:
Yes, I needed it to run it outside of ansible-test with my own profile so I could debug it

s-hertel commented on lib/ansible/modules/cloud/amazon/s3_bucket.py:
```diff
@@ -278,6 +285,9 @@ def create_or_update_bucket(s3_client, module, location):
             tags = dict((to_text(k), to_text(v)) for k, v in tags.items())
             if current_tags_dict != tags:
                 if tags:
+                    if not purge_tags:
```
Maybe this should be done above `if tags:` on line 287 since `tags: {}` and `purge_tags: False` could be used together.

willthames replied:
My thinking was that it would be weird to explicitly set `tags: {}` and `purge_tags: False` together - but it could happen when tags is a variable, so thanks for making me reconsider!

willthames replied:
Handling this and adding some extra tests around it uncovered some additional bugs, so this suggestion helped in more ways than one, thanks!

willthames committed:
s3_bucket: Add further tests and improve tag handling further

Additional tests for purge_tags: False suggested some incorrect
logic and thus further improvements

Increase wait timeout on bucket deletion as it wasn't always completing
in the default 100 seconds
